dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java

```diff
@@ -30,9 +32,39 @@ public void run(Program program) {
                 obj.getInitialValue(field).accept(collector);
             }
         }
-        // FIXME: We should also traverse the program spec for references to memory objects
+        // Traverse the program spec for references to memory objects
+        addSpecMemoryObjects(program.getSpecification(), collector);
         Sets.difference(memory.getObjects(), collector.memoryObjects).forEach(memory::deleteMemoryObject);
-         */
+    }
+
+    private void addSpecMemoryObjects(AbstractAssert spec, MemoryObjectCollector collector) {
+        if (spec instanceof AssertInline assertInline) {
+            addExprMemoryObject(assertInline.getAssertion().getExpression(), collector);
+        }
+        if (spec instanceof AssertBasic assertBasic) {
+            addExprMemoryObject(assertBasic.getLeft(), collector);
+            addExprMemoryObject(assertBasic.getRight(), collector);
+        }
```

Comment:
you can just do `expr.accept(collector)` in both cases here and the `addExprMemoryObject` method becomes obsolete.
However, you also need to overwrite `MemoryObjectCollector.visitLocation` (which could just visit the underlying memory object).

Reply:
Thanks!